### DISCUSSION TITLE
Storybook, NextJs,  swr, and mocking.

### DISCUSSION THREAD

**Discussion by hvgab** [storybookjs/storybook#26307] [Help] [answered]
### Summary

Hi all, I'm fairly new to both NextJs and Storybook, as far as I remember somewhere in the NextJs documentation they recommend you to put the api requests inside many components, and NextJs will deduplicate the requests.
Given that I then have a component that does this
(example:)
```ts
"use client";
import useSWR from "swr";
import fetcher from "../lib/fetcher";

export default function ServerPlayers({ serverId }: { serverId: string }) {
  const { data, error, isLoading } = useSWR(
    `/api/servers/${serverId}/info`,
    fetcher
  );

  if (isLoading)
    return (
      <span className="h-2.5 bg-gray-200 rounded-full dark:bg-gray-700 w-48 mb-4"></span>
    );

  if (error) return <span>Error Failed to fetch data</span>;

  return (
    <span>
      {data?.info?.players?.online}
      /
      {data?.info?.players?.max}
    </span>
  );
}
```
How should I do this inside Storybook, should I make a proxy for the api so I can show the components with real data, or should I somehow mock this data?
Given that the api request is inside the component, how would I then mock the data?
Or have I completely missed the point of Storybook, and should only use storybook for design? I was hoping I could write the component once, both for use in the project and for the Storybook.

**Comment by valentinpalkovic** [comment, marked as answer]
In this case, utilizing Mock Service Worker (MSW) alongside Storybook is a highly effective strategy for mocking your API requests directly within your components, including the scenario you described involving NextJs and useSWR for fetching data. MSW allows you to intercept and mock HTTP requests, which means you can simulate your API behavior directly in Storybook without the need for an actual backend during development. This not only enhances the efficiency of developing and testing components but also ensures a seamless integration with the Storybook environment.

To begin working with MSW in your Storybook setup, you would first install the MSW and the msw-storybook-addon by using npm or yarn commands as highlighted in the below mentioned documentation. This setup is straightforward and involves generating a service worker in your public folder, followed by configuring the addon within your Storybook's preview.js file to enable MSW.

Once MSW is configured in Storybook, you can start defining request handlers for your API endpoints. For the example you provided, where the component fetches server player information, you would mock the /api/servers/${serverId}/info endpoint. This is done by passing request handlers into the handlers property of the msw parameter within your story parameters, allowing you to specify the mock response you wish to return.

Furthermore, MSW's flexibility enables advanced usage like composing request handlers, inheriting, and selectively overriding or disabling global handlers directly in your stories. This means you can globally define common API behaviors and selectively adjust them for individual stories as needed [1](https://www.google.com/search?region=wt-wt&q=msw+Storybook+addon+for+mocking+API+requests&tbm=nws) .

Storybook indeed goes beyond merely serving as a tool for designing components. It enables developers to document, test, and visualize components in various states and interactions, including scenarios involving data fetching. Leveraging MSW with Storybook, as described, aligns perfectly with this objective by providing a controlled, mock backend environment. This approach ensures that you can develop, demonstrate, and test your components in isolation, including those that rely on API data, enhancing both the developer experience and the quality of the components.

For a detailed guide on installing, setting up, and configuring MSW in Storybook, you can refer to the official documentation available at https://storybook.js.org/addons/msw-storybook-addon . This resource provides comprehensive information to help you get started with mocking API requests in your Storybook projects effectively.

Please be aware, that the latest stable version of `msw-storybook-addon` only supports MSW 1.x. MSW 2.x was recently published, and the addon has experimental support for it, which is tracked here: https://github.com/mswjs/msw-storybook-addon/pull/122
